apps/events/api/permissions.py

```diff
@@ -0,0 +1,78 @@
+from django.utils import timezone
+from rest_framework import permissions
+
+from apps.authentication.models import OnlineUser as User
+
+from ..models import AttendanceEvent, Attendee
+
+
+class RegisterPermission(permissions.IsAuthenticated):
+    """
+    Check if a user has permission to register for an attendance event.
+    """
+
+    message = ""
+
+    def has_object_permission(self, request, view, obj: AttendanceEvent):
+        attend_response = obj.is_eligible_for_signup(request.user)
+        can_attend = attend_response.get("status")
+        self.message = attend_response.get("message")
+        return can_attend
+
+
+class UnregisterPermission(permissions.IsAuthenticated):
+    """
+    Check if a user has permission to unregister from an attendance event.
+    """
+
+    message = ""
+
+    @staticmethod
+    def _is_user_attending(attendance_event: AttendanceEvent, user: User):
+        return attendance_event.is_attendee(user)
+
+    @staticmethod
+    def _has_deadline_passed(attendance_event: AttendanceEvent, user: User) -> bool:
+        """ User can only be unattended before the deadline, or if they are on the wait list. """
+        now = timezone.now()
+        return (
+            now > attendance_event.unattend_deadline
+            and not attendance_event.is_on_waitlist(user)
+        )
```

Comment:
This permission is only used for the `unregsiter` endpoint, which is what a user would use to register.

Since registering people manually as an admin has a different set of constraints that should be another endpoint with another permission.

So yes, admins should be able to bypass this, but not on this endpoint. When registering on the regular page you are equal to everyone else, even when you are an admin